🖼 Integrate `jest-preview` into `react-hook-form`

nvh95 opened the pull request:
## Demo

![Preview Jest test in a browser](https://user-images.githubusercontent.com/8603085/175817421-c04108de-0469-48f3-9018-2eea38f2dd91.gif)

Visual debug for this test:

https://github.com/react-hook-form/react-hook-form/blob/36b5f5175c55d0c3b489cba0debef3ace3796bb0/src/__tests__/controller.test.tsx#L498-L546

## What is Jest Preview

- A library that previews the UI when we write Jest tests. For details document, please visit https://www.jest-preview.com/docs/getting-started/intro

## Why Jest Preview

- Provide a visual debugging experience to boost the productivity when writing Jest tests, which leads to:
  - Quickly identify the bug
  - Write a new test faster (by putting the `debug()` at the end of `it` block)

## Usage

- Start the Jest Preview Dashboard by `yarn jest-preview`.
- Just plug `debug()` wherever we want to preview the UI then run `yarn test:watch`.
- Opt-in to [Automatic Mode](https://www.jest-preview.com/blog/automatic-mode). Jest Preview automatically preview the UI whenever a Jest test fails.

## Others

- We need to [configure Jest Preview a little](https://www.jest-preview.com/docs/getting-started/installation#2-configure-jests-transform-to-transform-css-and-files) to see the CSS and images. But `react-hook-form` does not contain any CSS and images. So we can skip this one.

@bluebill1049 Please give me your thoughts on integrating `jest-preview` to `react-hook-form` to boost productivity when we develop features/ fix bugs. I attach an example in the source code. You just need to uncomment and follow the instruction

nvh95 committed:
🖼 feat: integrate jest-preview into react-hook-form

nvh95 committed:
🤖 feat: Opt-in to Automatic Mode

bluebill1049 commented:
Haha, I was going to ask for your help, and you already contributed to it.

bluebill1049 commented on src/__tests__/controller.test.tsx:
```diff
@@ -536,6 +537,12 @@ describe('Controller', () => {
     fireEvent.click(screen.getByRole('button', { name: /append/i }));
 
     expect(screen.getByRole('textbox')).toHaveValue('test');
+    // To see the actual UI in Jest, do the following:
+    // 1. Run `yarn jest-preview` to open Jest Preview Dashboard
+    // 2. Uncomment `debug()`
+    // 3. Run `yarn test:watch src/__tests__/controller.test.tsx`
+    // You should see the UI in your default browser.
+    // debug();
```
maybe move the instruction into the contribution.md?

nvh95 replied:
Sure. I updated.

nvh95 committed:
📖 docs: Update cypress commands in CONTRIBUTING.md

nvh95 committed:
📓 docs: Add usage instruction to visual debug using jest-preview

nvh95 committed:
⚙️ feat: configure css and file transform as well.
It wouldn't hurt to configure `jest-preview` properly just in case we have CSS and file in tests in the future

nvh95 commented:
@bluebill1049 Thanks for your feedback. I updated the code with the following changes:

1. Move instruction to visual debug Jest to `CONTRIBUTING.md`
2. Update instructions for running the automation test suite. The current instruction does not work since `yarn cypress` doesn't run if `yarn start:app` hasn't finished runnning.
3. Configure CSS and file transformations for `jest-preview`. Even though current tests don't it, in the future, if we use one of those, there won't be a surprising bug. Also, if someone read `react-hook-form` and sees `jest-preview` and want to copy the implementation, they would not miss some configurations.
Reference: [Jest Preview Installation documentation](https://www.jest-preview.com/docs/getting-started/installation#2-configure-jests-transform-to-transform-css-and-files)

bluebill1049 reviewed (approved):
Amazing stuff for DX 🥳 thank you again @nvh95

nvh95 commented:
Thank you very much @bluebill1049.

Upinwv commented:
Canary Mail
You've received a secure email
***@***.*** has sent you a secure email via Canary Mail.

Read Secure Email (https://secure.canarymail.io/read?obj_id=df1d7f79-94a9-4924-bf99-27324550b34b&obj_key=TDRrNU1LWk1MbVhnT1B5amVPK1ZRZz09&thr_id=df1d7f79-94a9-4924-bf99-27324550b34b)

If you expect to correspond often with ***@***.***, we recommend you download Canary Mail for free.

Download Canary (https://canarymail.io)

Privacy (https://canarymail.io/privacy.html) | Terms (https://canarymail.io/terms.html) | Docs (https://help.canarymail.io/) | Support (https://canarymail.zendesk.com/hc/en-us/requests/new)

Copyright © 2021 Canary Mail, All rights reserved.
